File: packages/setorra-ai/src/core/schema.ts
/**
 * Schema types for Setorra Node SDK (parity with Python v1.3).
 *
 * Additive-only: do not remove or rename fields once released.
 */

export const SCHEMA_VERSION = "1.3";

export interface EvidenceEvent {
  schema_version: string;
  session_id: string;
  event_index: number;
  timestamp: string; // ISO-8601 Z
  agent_name: string;
  agent_version: string;
  event_type: string;
  parameters_redacted: unknown;
  privacy_flags: string[];
  redacted_fields: string[];
  context: Record<string, unknown>;
  execution: Record<string, unknown>;
  policy: Record<string, unknown> | null;
  approval: Record<string, unknown> | null;
  guardrail: Record<string, unknown> | null;
  integrity: { prev_hash: string | null; event_hash: string; signature?: string | null };
}

export interface OutputSummary {
  schema_version: string;
  session_id: string;
  agent_name: string;
  agent_version: string;
  started_at: string;
  finished_at: string;
  duration_ms: number;
  model_id?: string | null;
  prompts: Record<string, { hash: string | null; redacted: string | null }>;
  final_output: unknown;
  run_result: Record<string, unknown>;
  tools_used: Array<Record<string, unknown>>;
  reasoning: Array<Record<string, unknown>>;
  consents: Array<Record<string, unknown>>;
  latency_ms?: number | null;
  cost_estimate?: number | null;
  environment_info: Record<string, unknown>;
  policy_summary: Record<string, unknown>;
  guardrail_summary: Record<string, unknown>;
  actions?: {
    total: number;
    by_status: Record<string, number>;
    final_statuses: Array<Record<string, unknown>>;
  };
  integrity: { first_event_hash: string | null; last_event_hash: string | null; count: number; signature?: string | null };
  organization?: { org_id: string; org_name?: string; key_id_prefix?: string };
}

File: setorra/schema.py
"""Shared schema types and helpers for Setorra.

Schema versions are additive-only. Keep field names stable and avoid breaking changes.
"""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import Any, Dict, List, Optional


# Additive-only schema bump for planning capture events and optional plan
# context on events. Backwards-compatible.
SCHEMA_VERSION = "1.3"


@dataclass
class EvidenceEvent:
    schema_version: str
    session_id: str
    event_index: int
    timestamp: str
    agent: Dict[str, str]
    event_type: str
    parameters_redacted: Any
    privacy_flags: List[str]
    redacted_fields: List[str]
    context: Optional[Dict[str, Any]] = None
    execution: Optional[Dict[str, Any]] = None
    policy: Optional[Dict[str, Any]] = None
    approval: Optional[Dict[str, Any]] = None
    guardrail: Optional[Dict[str, Any]] = None
    integrity: Dict[str, Any] = field(default_factory=dict)


@dataclass
class OutputSummary:
    schema_version: str
    session_id: str
    timestamp: str
    content: str
    prompts: Optional[Dict[str, Any]] = None
    tools_used: List[Dict[str, Any]] = field(default_factory=list)
    errors: List[Dict[str, Any]] = field(default_factory=list)
    integrity: Dict[str, Any] = field(default_factory=dict)
